crates/runtime/src/actor/component.rs

```diff
@@ -0,0 +1,863 @@
+        let bindings = self.as_guest_bindings().await?;
+        Ok(GuestInstance {
+            store: Arc::new(Mutex::new(self.store)),
+            bindings: Arc::new(bindings),
+        })
+    }
+
+    /// Instantiates and returns a [`InterfaceInstance<incoming_http_bindings::IncomingHttp>`] if exported by the [`Instance`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if incoming HTTP bindings are not exported by the [`Instance`]
+    pub async fn into_incoming_http(
+        mut self,
+    ) -> anyhow::Result<InterfaceInstance<incoming_http_bindings::IncomingHttp>> {
+        let bindings = if let Ok((bindings, _)) =
+            incoming_http_bindings::IncomingHttp::instantiate_async(
+                &mut self.store,
+                &self.component,
+                &self.linker,
+            )
+            .await
+        {
+            InterfaceBindings::Interface(bindings)
+        } else {
+            self.as_guest_bindings()
+                .await
+                .map(InterfaceBindings::Guest)
+                .context("failed to instantiate `wasi:http/incoming-handler` interface")?
+        };
+        Ok(InterfaceInstance {
+            store: Mutex::new(self.store),
+            bindings,
+        })
+    }
+
+    /// Instantiates and returns an [`InterfaceInstance<logging_bindings::Logging>`] if exported by the [`Instance`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if logging bindings are not exported by the [`Instance`]
+    pub async fn into_logging(
+        mut self,
+    ) -> anyhow::Result<InterfaceInstance<logging_bindings::Logging>> {
+        let bindings = if let Ok((bindings, _)) = logging_bindings::Logging::instantiate_async(
+            &mut self.store,
+            &self.component,
+            &self.linker,
+        )
+        .await
+        {
+            InterfaceBindings::Interface(bindings)
+        } else {
+            self.as_guest_bindings()
+                .await
+                .map(InterfaceBindings::Guest)
+                .context("failed to instantiate `wasi:logging/logging` interface")?
+        };
+        Ok(InterfaceInstance {
+            store: Mutex::new(self.store),
+            bindings,
+        })
+    }
+}
+
+enum GuestBindings {
+    Command(Command),
+    Interface(guest_bindings::Guest),
+}
+
+impl GuestBindings {
+    /// Invoke an operation on a [GuestBindings] producing a result.
+    #[instrument(skip_all)]
+    pub async fn call(
+        &self,
+        mut store: &mut wasmtime::Store<Ctx>,
+        operation: impl AsRef<str>,
+        request: impl AsyncRead + Send + Sync + Unpin + 'static,
+        response: impl AsyncWrite + Send + Sync + Unpin + 'static,
+    ) -> anyhow::Result<Result<(), String>> {
+        let ctx = store.data_mut();
+        ctx.stdin.replace(Box::new(request)).await;
+        ctx.stdout.replace(Box::new(response)).await;
+        let res = match self {
+            GuestBindings::Command(bindings) => {
```

Comment:
Making sure I grok: so `Command` will be used for Go actors, and `Interface` for Rust?

Reply:
I'd describe this more as Command is for any WASI Command style module/component. GoLang is the only language that I know of that _only_ supports wasi preview 1 command style modules, but theoretically lots of folks may choose to write command style (aka with a main func) wasi module/component/actor.